kong/plugins/bot-detection/handler.lua

```diff
@@ -0,0 +1,71 @@
+local BasePlugin = require "kong.plugins.base_plugin"
+local reg = require "rex_pcre"
+local responses = require "kong.tools.responses"
+local rules = require "kong.plugins.bot-detection.rules"
+local stringy = require "stringy"
+local bot_cache = require "kong.plugins.bot-detection.cache"
+
+local ipairs = ipairs
+
+local BotDetectionHandler = BasePlugin:extend()
+
+BotDetectionHandler.PRIORITY = 2500
+
+local ALLOW = 1
+local BLOCK = 2
+
+function BotDetectionHandler:new()
+  BotDetectionHandler.super.new(self, "bot-detection")
+end
+
+function BotDetectionHandler:access(conf)
+  BotDetectionHandler.super.access(self)
+
+  local user_agent = ngx.req.get_headers()["user-agent"]
+  if user_agent then
+    user_agent = stringy.strip(user_agent)
+
+    -- Cache key, per API
+    local cache_key = ngx.ctx.api.id..":"..user_agent
+
+    -- The cache already has the user_agents that should be blocked
+    -- So we avoid matching the regexes everytime
+    local cached_match = bot_cache.get(cache_key)
+    if cached_match and cached_match == ALLOW then
+      return
+    elseif cached_match and cached_match == BLOCK then
+      return responses.send_HTTP_FORBIDDEN()
+    end
+
+    if conf then
```

Comment:
This is not needed, the plugin execution guarantees `conf` will be a table, at the very least an empty one, iirc. This is on purpose to avoid having such `if conf then` tests in the plugin handlers themselves.